dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java

```diff
@@ -17,156 +17,342 @@
+    // @formatter:off
+    //      {
+    //        "fields": [
+    //          "donor_id",
+    //          "specimen_id"
+    //        ],
+    //        "bidirectional": false,
+    //        "other": "specimen",
+    //        "otherFields": [
+    //          "donor_id",
+    //          "specimen_id"
+    //        ],
+    //        "optionals": [
+    //          1
+    //        ]
+    //      }
+    // @formatter:on
+    // resolveForeignKeys() and resolveOptionalKeys() take into consideration optionals when resolve keys
+    val foreignKeys = resolveForeignKeys(schemaFieldNames, relations, fileSchema);
+    val optionalKeys = resolveOptionalKeys(schemaFieldNames, relations, fileSchema);
+    val conditionalKeys = resolveConditionalKeys(schemaFieldNames, relations, fileSchema);
+
+    builder.put(FK, foreignKeys);
+    builder.put(OPTIONAL_FK, optionalKeys);
+    builder.put(CONDITIONAL_FK, conditionalKeys);
+
+    return new KVFileTypeKeysIndices(builder.build(), createConditionEvaluators(fileType));
   }
 
   @Override
-  public List<KVFileType> getExperimentalFileTypes(KVExperimentalDataType dataType) {
-    return copyOf(transform(
-        dictionary.getFileTypesReferencedBranch(dataType.getFeatureType()),
-        new Function<FileType, KVFileType>() {
-
-          @Override
-          public KVFileType apply(FileType fileType) {
-            return KVFileType.from(fileType);
-          }
-
-        }));
-  }
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#hasOutgoingSurjectiveRelation(org.icgc.dcc.submission.
-   * validation.key.core.KVFileType)
-   */
-  @Override
-  public boolean hasOutgoingSurjectiveRelation(KVFileType fileType) {
-    // TODO Auto-generated method stub
-    return false;
-  }
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getKeysIndices(org.icgc.dcc.submission.validation.key.
-   * core.KVFileType)
-   */
-  @Override
-  public KVFileTypeKeysIndices getKeysIndices(KVFileType fileType) {
-    // TODO Auto-generated method stub
-    return null;
-  }
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getOptionalReferencedFileType1(org.icgc.dcc.submission
-   * .validation.key.core.KVFileType)
-   */
-  @Override
-  public Optional<KVFileType> getOptionalReferencedFileType1(KVFileType fileType) {
-    // TODO Auto-generated method stub
-    return null;
-  }
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getOptionalReferencedFileType2(org.icgc.dcc.submission
-   * .validation.key.core.KVFileType)
-   */
+  public List<String> getErrorFieldNames(@NonNull KVFileType fileType, @NonNull KVErrorType errorType,
+      @NonNull Optional<KVFileType> optionalReferencedFileType) {
```

Comment:
With respect to `Optional` parameters, something to consider:

http://blog.jhades.org/java-8-how-to-use-optional/
http://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments